### PR TITLE
Speed up DesignSpace 

### PR DESCRIPTION
**Pull request by Paul-Saves**
Speed up DesignSpace verifications and rounding for the case in which the verifications and rounding do not apply to save time.
To avoid systematic verification many if-else clause and boolean have been added. 
Linting and formatting with ruff 3.7.

### CODE REVIEW

File: smt/surrogate_models/krg_based.py [hunk 1]
```diff
@@ -1463,6 +1463,8 @@ def _predict_values(self, x: np.ndarray, is_acting=None) -> np.ndarray:
             X_cont = (X_cont - self.X_offset) / self.X_scale
 
         else:
+            n_eval, n_features_x = x.shape
```
Comment (relf) [inline smt/surrogate_models/krg_based.py:1466]: `n_feature_x` does not seem to be used below, either replace by `_` or this line could stay before the `if`

File: smt/surrogate_models/krg_based.py [hunk 2]
```diff
@@ -1463,6 +1463,8 @@ def _predict_values(self, x: np.ndarray, is_acting=None) -> np.ndarray:
...
+            _, ij = cross_distances(x, self.X_train)
```
Comment (relf) [inline smt/surrogate_models/krg_based.py:1467]: `ij` does not seem to used below and if `cross_distances` has no side-effect (what I hope!), you could remove this line.

File: smt/surrogate_models/krg_based.py [hunk 3]
```diff
@@ -1647,6 +1649,8 @@ def _predict_variances(self, x: np.ndarray, is_acting=None) -> np.ndarray:
             X_cont, _ = compute_X_cont(x, self.design_space)
             X_cont = (X_cont - self.X_offset) / self.X_scale
         else:
+            n_eval, n_features_x = x.shape
+            _, ij = cross_distances(x, self.X_train)
```
Comment (relf) [inline smt/surrogate_models/krg_based.py:1653]: Same comments as above. Actually, you should create a function to remove code duplication.

**Review by relf — CHANGES_REQUESTED** [review]

File: smt/surrogate_models/krg_based.py (hunk 3 above)
Comment (relf) [reply smt/surrogate_models/krg_based.py]: Not solved for me. Could you extract common init code between `_predict_values` and `_predict_variances` in a function `_predict_init()`? In order to replace code duplication by:
```
r, X_cont = self._predict_init(self, ...)
```

File: smt/utils/design_space.py [hunk 4]
```diff
@@ -894,10 +895,16 @@ def add_value_constraint(
...
         # Get parameters
         param1 = self._get_param(var1)
         param2 = self._get_param(var2)
+        if (
+            isinstance(param1, UniformIntegerHyperparameter)
+            or isinstance(param2, UniformIntegerHyperparameter)
+            or isinstance(param1, OrdinalHyperparameter)
+            or isinstance(param2, OrdinalHyperparameter)
+        ):
+            self.has_valcons_ord_int = True
```
Comment (relf) [inline smt/utils/design_space.py:907]: You could write
```
            mixint_types = (UniformIntegerHyperparameter, OrdinalHyperparameter)
            self.has_valcons_ord_int = isinstance(param1, mixint_types) or isinstance(param2, mixint_types)
```

File: smt/applications/mixed_integer.py [hunk 5]
```diff
@@ -224,9 +224,9 @@ def __init__(
             )
             and self._surrogate.options["categorical_kernel"] is None
         ):
-            self._surrogate.options["categorical_kernel"] = (
-                MixIntKernelType.HOMO_HSPHERE
-            )
+            self._surrogate.options[
+                "categorical_kernel"
+            ] = MixIntKernelType.HOMO_HSPHERE
```
Comment (relf) [inline smt/applications/mixed_integer.py:229]: Why is this reformatted? Do you still use black?

File: smt/utils/neural_net/model.py [hunk 6]
```diff
@@ -221,9 +221,9 @@ def train(
 
                 # Compute average cost and print output
                 avg_cost = np.mean(optimizer.cost_history).squeeze()
-                self._training_history["epoch_" + str(e)]["batch_" + str(b)] = (
-                    optimizer.cost_history
-                )
+                self._training_history["epoch_" + str(e)][
+                    "batch_" + str(b)
+                ] = optimizer.cost_history
```
Comment (relf) [inline smt/utils/neural_net/model.py:226]: black again?

Comment (Paul-Saves) [reply smt/utils/neural_net/model.py:226]: no "ruff format"

File: smt/applications/mixed_integer.py (hunk 5 above)
Comment (Paul-Saves) [reply smt/applications/mixed_integer.py:229]: no, ruff format

Comment (Paul-Saves) [reply smt/applications/mixed_integer.py:229]: ruff has been updated to 3.6 last week and to 3.7 this week.

File: smt/utils/neural_net/model.py (hunk 6 above)
Comment (Paul-Saves) [reply smt/utils/neural_net/model.py:226]: ruff has been updated to 3.6 last week and to 3.7 this week.

Comment (Paul-Saves) [reply smt/utils/neural_net/model.py:226]: ruff has been updated to 3.6 last week and to 3.7 this week.

File: smt/applications/mixed_integer.py (hunk 5 above)
Comment (Paul-Saves) [reply smt/applications/mixed_integer.py:229]: ruff has been updated to 3.6 last week and to 3.7 this week.

**Review by relf — APPROVED** [review]

**Pull request merged**